sigopt_sklearn/search.py

```diff
@@ -175,46 +179,58 @@ def __init__(self, estimator, param_domains, n_iter=10, scoring=None,
-            any_floats = any(isinstance(x, float) for x in bounds)
-            all_strings = all(isinstance(x, str) for x in bounds)
-
-            # convert entire bounds to floats if one is present
-            if any_floats:
-                bounds = [float(b) for b in bounds]
-                all_floats = True
-
-            types = ['double', 'int', 'categorical']
-            bound_types = [all_floats, all_ints, all_strings]
-            param_type = types[bound_types.index(True)]
-
-            param_dict = {}
-            param_dict['name'] = param_name
-            param_dict['type'] = param_type
-
-            if param_type == 'categorical':
-                cat_vals = []
-                for str_name in bounds:
-                    cat_vals.append({'name': str_name})
-                param_dict['categorical_values'] = cat_vals
+        def _transform_param(param_name, param_bounds):
+            """Transform a parameter name and its bounds into a form that can be sent to the API layer."""
+            def _check_bounds():
+                """Check that min/max bounds are well formed."""
+                if len(param_bounds) != 2:
+                    raise Exception('Parameter bounds must be specified with two numbers! Not sure what to do with {}.'
+                                    .format(param_bounds))
+                if not isinstance(param_bounds, tuple):
+                    warnings.warn('Parameter bounds should be specified as a tuple in the form (min, max).')
+
+            param_dict = {'name': param_name}
+            if isinstance(param_bounds, dict):
+                # This is a categorical with mappings between strings and values
+                param_dict['type'] = 'categorical'
+                param_dict['categorical_values'] = [{'name': k} for k in param_bounds.keys()]
+
+                # Add this mapping to our set of categorical string mappings
+                self.categorical_mappings_[param_name] = param_bounds
+
+            elif all(isinstance(x, str) for x in param_bounds):
+                # This is a categorical with a list of strings naming each category
+                param_dict['type'] = 'categorical'
+                param_dict['categorical_values'] = [{'name': k} for k in param_bounds]
+
+            elif all(isinstance(x, int) for x in param_bounds):
+                # This is an integer parameter
+                _check_bounds()
+                param_dict['type'] = 'int'
+                param_dict['bounds'] = {'min': param_bounds[0], 'max': param_bounds[1]}
+
+            elif any(isinstance(x, float) for x in param_bounds):
+                # This is a continuous parameter. Note that we use `any` since the user may pass some combination of
+                # float and integer parameters, e.g. (0, 0.1).
+                _check_bounds()
+                param_dict['type'] = 'double'
+                param_dict['bounds'] = {'min': param_bounds[0], 'max': param_bounds[1]}
```

Comment:
could we add some fail through error here that might help someone diagnose that they've specified the parameter domains in a weird way?  Love that you've added other exceptions though +1 on that